js/stdlib/ts/shared_impl.ts

```diff
@@ -200,6 +209,28 @@ export type ISetupView<ContractInfo, VerifyResult, ConnectorTy extends AnyBacken
   getView1: ((views:IBackendViewsInfo<ConnectorTy>, v:string, k:string|undefined, vi:IBackendViewInfo<ConnectorTy>, isSafe: boolean) => ViewVal)
 };
 
+export type ISetupEvent<ContractInfo, VerifyResult, ConnectorTy extends AnyBackendTy> = (args:ISetupEventArgs<ContractInfo, VerifyResult>, bin: IBackend<ConnectorTy>) => {
+  getEvent: (name: string, ty: any) => Promise<Event<any>>;
+}
+
+export type Time = number;
+
+export type Event<T> = {
+  when: Time,
+  what: T
+}
+
+export type EventStream<T> = {
+  // mvp
+  seek: (t:Time) => void,
+  next: () => Promise<Event<T>>
+  // additional
+  seekNow: () => void,
+  lastTime: () => Time,
+  // why can't TS handle a function type as arg
```

Comment:
You just need lots of parens

Reply:
`monitor: (fn: () => void) => void` is valid TypeScript. Depending on the function signature of `monitor`, you might want to change that.

For instance, if `monitor` takes a sum function that takes two `number` parameters `a` and `b`, and this sum function returns their sum, also a `number`, and `monitor` returns `void`, we could write that as `monitor: (fn: (a: number, b: number) => number) => void`.

Reply:
`monitor: (fn: () => void) => void` is valid TypeScript. Depending on the function signature of `monitor`, you might want to change that.

For instance, if `monitor` takes a sum function that takes two `number` parameters `a` and `b`, and this sum function returns their sum, also a `number`, and `monitor` returns `void`, we could write that as

```
monitor: (fn: (a: number, b: number) => number) => void
```